src/main/scala/services/ReservationCreateService.scala

```diff
@@ -0,0 +1,142 @@
+    val pattern = "^[A-Z][a-ząćęłńóśźżA-ZĄĆĘŁŃÓŚŹŻ]{2,}$".r
+    pattern.matches(value)
+  }
+
+  private def isValidSurname(value: String): Boolean = {
+    val pattern = "^[A-Z][a-ząćęłńóśźżA-ZĄĆĘŁŃÓŚŹŻ]{2,}(-[A-Z][a-ząćęłńóśźżA-ZĄĆĘŁŃÓŚŹŻ]{2,})?$".r
+    pattern.matches(value)
+  }
+
+  override def createReservation(reservationRequest: CreateReservationRequest): Future[ReservationCreateResult] = {
+    logger.info(s"Creating reservation for: ${reservationRequest.firstName} ${reservationRequest.lastName}")
+    validateReservationRequest(reservationRequest) match {
+      case Some(error) => Future.successful(error)
+      case None        => processReservation(reservationRequest)
+    }
+  }
+
+  private def validateReservationRequest(reservationRequest: CreateReservationRequest): Option[ReservationCreateResult] = {
+    if (!isValidName(reservationRequest.firstName) || !isValidSurname(reservationRequest.lastName))
+      Some(InvalidFirstNameOrLastNameFormat)
+    else if (reservationRequest.selectedSeatsWithTicketsTypes.isEmpty)
+      Some(NoSeatsSelected)
+    else
+      None
+  }
+
+  private def processReservation(reservationRequest: CreateReservationRequest): Future[ReservationCreateResult] = {
+    reservationRepository
+      .validateSeatSelection(
+        reservationRequest.selectedSeatsWithTicketsTypes.map(_.seatId),
+        reservationRequest.screeningId
+      )
+      .getOrElse(false)
+      .flatMap { isValid =>
+        if (isValid) createAndSaveReservation(reservationRequest)
+        else         Future.successful(InvalidSeatSelection)
+      }
+  }
+
+  private def createAndSaveReservation(reservationRequest: CreateReservationRequest): Future[ReservationCreateResult] = {
+    val isReservationPossibleF = checkIfReservationIsPossibleByTime(reservationRequest.screeningId).getOrElse(false)
+
+    isReservationPossibleF.flatMap { isReservationPossible =>
+      if (isReservationPossible) {
+        val reservation = createReservationObject(reservationRequest)
+        val reservedSeats = createReservedSeatsObjects(reservationRequest, reservation.id)
+        val reservationInfo = ReservationListService.createReservationInfo(reservation, reservedSeats)
+
+        for {
+          _ <- reservationRepository.createReservation(reservation, reservedSeats)
+        } yield Created(reservationInfo)
+      } else
+        Future.successful(ReservationNotPossible)
+    }
+  }
+
+  private def checkIfReservationIsPossibleByTime(screeningId: UUID): OptionT[Future, Boolean] = {
+    val currentTime    = LocalDateTime.now()
+    val screening      = screeningRepository.getScreening(screeningId)
+    val screeningTime  = screening.map(_.startTime)
+    val timeDifference = screeningTime.map(_.minusMinutes(15))
+
+    for {
+      timeDifference <- timeDifference
+    } yield {
+      if(currentTime.isBefore(timeDifference)) {
+        true
+      } else {
+        false
+      }
+    }
+  }
```

Comment:
`isBefore` zwraca już `Boolean` więc nie musimy explicit robić `if(true) true else false`. Można również zastanowić się czy używanie for comprehension dla jednej wartości i bez warunków (filtrów) jest czytelniejsze niż alternatywa w postaci ... :)

Reply:
Ahaa, taak. Można to znacziej uprościć :)